preview thumbnail in video progress bar / preview on video scrubbing

rigelk opened the pull request:
![Screenshot from 2019-12-08 15-01-38](https://user-images.githubusercontent.com/6329880/70390763-4af8d400-19ce-11ea-81ce-5587102b6605.png)

The PR generates extra images in the post-upload phase of a video, as well as a `text/vtt` file that tracks ranges to which the preview correspond. It relies on a plugin for videojs, https://www.npmjs.com/package/videojs-vtt-thumbnails, which I adapted to provide functional previews in the `pthumbnail` package (yes, all plugins are more or less broken).

Known WIPs (apart from typos and quick code) (checkmark == resolved):
- [ ] federation support
- [x] ensure function works for embeds
- [x] image generation slows noticeably the post-upload phase (~= a few seconds)
- [x] synchronous fs check at the AP video representation - maybe I should add a video attribute in db instead?

The PR solves #537.

tcitworld commented:
> image generation slows noticeably the post-upload phase (~= a few seconds)

It will definitely be more than a few seconds for a whole movie, it should be fully async in that case.

rigelk commented:
> > image generation slows noticeably the post-upload phase (~= a few seconds)
> 
> It will definitely be more than a few seconds for a whole movie, it should be fully async in that case.

Considering longer videos don't get more thumbnails (rather, videos of duration smaller than the number of thumbnails set will be capped to 1 thumbnail/sec). So if I set for instance 100 thumbnails to generate for a video, generating images for a movie or a 100-sec-long clip will take the same time.

rigelk commented:
@Chocobozzz should I reuse the existing `VideoModel.Thumbnails` somehow, or should I add a field to point to the VTT file?

Chocobozzz commented on server/helpers/ffmpeg-utils.ts:
```diff
@@ -124,36 +125,71 @@ function getDurationFromVideoFile (path: string) {
+  if (duration && count > duration) count = duration
+
+  // pending images will have names ending in '-001', '-002', etc.
+  const insertBeforeExtension = (s, text) => s.substring(0, s.lastIndexOf('.')) + text + s.substring(s.lastIndexOf('.'))
```
Please create a dedicated function with comments

Chocobozzz commented on server/helpers/ffmpeg-utils.ts:
```diff
@@ -124,36 +125,71 @@ function getDurationFromVideoFile (path: string) {
+    width: number,
+    height: number
+  },
+  count = 1 // count > 1 generates thumbnails and their vtt file
```
I think it would be better to have a dedicated function instead of adding this parameter.

Chocobozzz commented on server/lib/thumbnail.ts:
```diff
@@ -61,7 +61,10 @@ function generateVideoMiniature (video: MVideoThumbnail, videoFile: MVideoFile,
-    : () => generateImageFromVideoFile(input, basePath, filename, { height, width })
+    : () => Promise.all([
+      generateImageFromVideoFile(input, basePath, filename, { height, width }),
+      generateImageFromVideoFile(input, basePath, filename, { height, width }, 10)
```
Should be async since the timestamp thumbnails are optional

rigelk replied:
Should I change something there?

Chocobozzz replied:
You could remove the second line, and call the function directly from the controller:

```
...
await generateImageFromVideoFile(...)
generateTimecodeThumbnails(...).catch(err => logger.error)
...
```

Chocobozzz commented on server/models/video/video-format-utils.ts:
```diff
@@ -303,6 +306,15 @@ function videoModelToActivityPubObject (video: MVideoAP): VideoTorrentObject {
   })
 
+  // Add video thumbnails, if any
+  if (pathExistsSync(join(CONFIG.STORAGE.THUMBNAILS_DIR, video.uuid + '.vtt'))) {
```
Never use sync.

Chocobozzz commented:
> @Chocobozzz should I reuse the existing VideoModel.Thumbnails somehow, or should I add a field to point to the VTT file?

You'll have to create a dedicated table to store the VTT information, `filename` (generated with an UUID) and `videoId`. We need the `filename` column to generate VTT files with a random UUID and then easily invalidate the cache by changing this name if the VTT file changes. We did not use this method for the thumbnails before so we are unable to put an important cache for these resources.

With this new table,  and an optional join you'll be able to remove your sync call in the format function by just checking if your VTT object exists.

fflorent commented:
This would fix https://github.com/Chocobozzz/PeerTube/issues/537

Florent

Chocobozzz commented on client/src/app/shared/video/video.model.ts:
```diff
@@ -96,6 +98,8 @@ export class Video implements VideoServerModel {
     this.thumbnailUrl = absoluteAPIUrl + hash.thumbnailPath
     this.previewPath = hash.previewPath
     this.previewUrl = absoluteAPIUrl + hash.previewPath
+    this.manifestPath = hash.manifestPath
```
Why not `vttThumbnailsPath`?

Chocobozzz commented on server/controllers/api/videos/index.ts:
```diff
@@ -223,13 +223,13 @@ async function addVideo (req: express.Request, res: express.Response) {
     : await generateVideoMiniature(video, videoFile, ThumbnailType.MINIATURE)
 
-  // Process preview or create it from the video
+  // Process preview or create it from the video (large thumbnail used for player backgrounds)
```
:+1:

Chocobozzz commented on server/controllers/api/videos/index.ts:
```diff
@@ -246,6 +246,19 @@ async function addVideo (req: express.Request, res: express.Response) {
 
+    // Process timecoded thumbnails from the video (small thumbnails bound to a timecode range, used for player scrubbing preview)
+    // We do not wait for it to finish as that non-critical task could impede the task length
+    generateVideoTimecodeThumbnails(video, videoFile, ThumbnailType.TIMECODE)
```
Put this block outside the first transaction

Chocobozzz commented on server/controllers/api/videos/index.ts:
```diff
@@ -246,6 +246,19 @@ async function addVideo (req: express.Request, res: express.Response) {
+      .then(res => {
+        sequelizeTypescript.transaction(async t2 => {
+          res.thumbnails.forEach(thumbnail => {
+            videoCreated.addAndSaveThumbnail(thumbnail, t2)
```
Use `await` (or `await Promise.all`

Chocobozzz commented on server/controllers/api/videos/index.ts:
```diff
@@ -246,6 +246,19 @@ async function addVideo (req: express.Request, res: express.Response) {
+          res.thumbnails.forEach(thumbnail => {
+            videoCreated.addAndSaveThumbnail(thumbnail, t2)
+          })
+          videoCreated.addAndSaveTimecodeThumbnailManifest(res.manifest, t2)
```
`await`

Chocobozzz commented on server/controllers/api/videos/index.ts:
```diff
@@ -246,6 +246,19 @@ async function addVideo (req: express.Request, res: express.Response) {
+          videoCreated.addAndSaveTimecodeThumbnailManifest(res.manifest, t2)
+        })
+      })
+      .catch(err => logger.error(err))
```
`logger.error('Cannot create timecode thumbnails for %s', video.url, { err })`

Chocobozzz commented on server/helpers/ffmpeg-utils.ts:
```diff
@@ -156,6 +169,104 @@ async function generateImageFromVideoFile (fromPath: string, folder: string, ima
+      for (let i = 1; i <= count; i++) {
+        const pendingImagePath = join(folder, pendingImageName(i))
+        await remove(pendingImagePath)
+      }
```
I think you need to exit the function here no?

Chocobozzz commented on server/helpers/ffmpeg-utils.ts:
```diff
@@ -156,6 +169,104 @@ async function generateImageFromVideoFile (fromPath: string, folder: string, ima
+  let duration = await getDurationFromVideoFile(fromPath)
+  if (count > duration) count = duration
+
+  imageName = TimecodeThumbnailManifestModel.generateManifestName().replace(/\.[^.]+$/, extname(imageName))
```
Please create a dedicated function to generate the image name, without a `replace`

Chocobozzz commented on server/helpers/ffmpeg-utils.ts:
```diff
@@ -156,6 +169,104 @@ async function generateImageFromVideoFile (fromPath: string, folder: string, ima
+
+  const manifest = new TimecodeThumbnailManifestModel()
+
+  manifest.filename = imageName.replace(/\.[^.]+$/, '.vtt')
```
You don't use `generateManifestName`?

Chocobozzz commented on server/helpers/ffmpeg-utils.ts:
```diff
@@ -156,6 +169,104 @@ async function generateImageFromVideoFile (fromPath: string, folder: string, ima
+    for (let i = 1; i <= count; i++) {
+      const thumbnailCreator = async () => {
+        const pendingImagePath = join(folder, pendingImageName(i))
+        const destination = join(folder, pendingImageName(i).replace('pending-', ''))
```
Please don't use `replace`

Chocobozzz commented on server/helpers/ffmpeg-utils.ts:
```diff
@@ -156,6 +169,104 @@ async function generateImageFromVideoFile (fromPath: string, folder: string, ima
+      identifier: '',
+      start: (i * duration / count),
+      end: ((i + 1) * duration / count),
+      text: join(STATIC_PATHS.THUMBNAILS, pendingImageName(i + 1).replace('pending-', '')),
```
Please don't use `replace`

Chocobozzz commented on server/models/video/thumbnail.ts:
```diff
@@ -72,6 +73,18 @@ export class ThumbnailModel extends Model<ThumbnailModel> {
 
+  @ForeignKey(() => TimecodeThumbnailManifestModel)
+  @Column
+  manifestId: number
```
`timecodeThumbnailManifestId`

I know it's long but I don't want to rename a foreign key in the future ^^

Chocobozzz commented on server/models/video/thumbnail.ts:
```diff
@@ -72,6 +73,18 @@ export class ThumbnailModel extends Model<ThumbnailModel> {
+    },
+    onDelete: 'CASCADE'
+  })
+  Manifest: TimecodeThumbnailManifestModel
```
Same here

Chocobozzz commented on package.json:
```diff
@@ -144,6 +146,7 @@
     "redis": "^2.8.0",
     "reflect-metadata": "^0.1.12",
     "request": "^2.81.0",
+    "rimraf": "^3.0.0",
```
it's unused

Chocobozzz commented on server/models/video/timecode-thumbnail-manifest.ts:
```diff
@@ -0,0 +1,90 @@
+    logger.info('Removing timecode thumbnail manifest file %s.', instance.filename)
+
+    // Don't block the transaction
+    instance.removeManifest()
```
Where do you remove timecode thumbnails?

Chocobozzz commented on server/models/video/timecode-thumbnail-manifest.ts:
```diff
@@ -0,0 +1,90 @@
+    }
+  ]
+})
+export class TimecodeThumbnailManifestModel extends Model<TimecodeThumbnailManifestModel> {
```
I think you should store how many thumbnails you generated

rigelk replied:
Done in the new version.

rigelk commented:
After looking at media.ccc, I noticed they added support for thumbnails scrubbing since the last time I checked (ages ago!) and they use the best of both worlds: a few thumbnail images (I counted 15 in [this example video]()), each actually being a sprite image weighting no more than 300kB so as to load as many images as possible in a given network call: https://static.media.ccc.de/media/congress/2019/10652-hd.thumbnails-09.jpg

The good news is, the videojs plugin I use already supports using sprites as image source, so I'll just need to modify the image generation to generate sprites.

rigelk commented:
@Chocobozzz I rebased and added most of the suggestions you made. I still need to achieve putting thumbnails in a subdirectory corresponding to the video uuid, though.

I have question, related to `server/typings/models/video/video.ts`. I added `MTimecodeThumbnailManifest` to the definitons of `MVideoAP` and `MVideoDetails` but am unsure if I should put it in other definitons, and what the side effects can be.

Chocobozzz commented on server/models/video/video.ts:
```diff
@@ -1572,8 +1598,7 @@ export class VideoModel extends Model<VideoModel> {
       ScopeNames.WITH_WEBTORRENT_FILES,
-      ScopeNames.WITH_STREAMING_PLAYLISTS,
-      ScopeNames.WITH_THUMBNAILS
+      ScopeNames.WITH_STREAMING_PLAYLISTS
```
You need the thumbnail here, but you can remove it from the condition below

Chocobozzz commented:
> Q1: related to server/typings/models/video/video.ts. I added MTimecodeThumbnailManifest to the definitons of MVideoAP and MVideoDetails but am unsure if I should put it in other definitons, and what the side effects can be.

You should put this definition in the models you need. Don't forget to adapt the SQL queries, to fetch `TimecodeThumbnailManifest` from the database when you build an `MVideoAP` and `MVideoDetails` object

> Q2: I have yet to decide how to preload timecode thumbnails. Right now it's all or nothing. Any preference? (naive approach could be neighbouring thumbnails after first hover, or preload mean representative thumbnails that would be displayed on a closest-cached basis until the timecode thumbnail is fetched). I'm wondering if sprite images doesn't help with the first scenario too.

I like the first approach. If you think sprites are simpler, you can generate them and lazy load them from remote instances.

> Troubleshooting: I have some trouble with the miniature, though. I don't know what change could have impacted it (a scope error?) but the error happens at the AP generation step: Cannot read property 'getFileUrl' of undefined\n at Object.videoModelToActivityPubObject (peertube/server/models/video/video-format-utils.ts:351:22). The undefined her is the miniature.getFileUrl line 351.

I added a comment that I think will fix your issue

rigelk commented on client/src/assets/player/peertube-player-manager.ts:
```diff
@@ -19,6 +20,7 @@ import { segmentValidatorFactory } from './p2p-media-loader/segment-validator'
 import { segmentUrlBuilderFactory } from './p2p-media-loader/segment-url-builder'
 import { RedundancyUrlManager } from './p2p-media-loader/redundancy-url-manager'
 import { getStoredP2PEnabled } from './peertube-player-local-storage'
+import { environment } from '../../environments/environment'
```
What should I use instead?

kimsible commented:
Any help to rebase this branch @rigelk ?

rigelk commented:
@kimsible I was planning to rebase the PR once the roadmap is complete, or maybe in september. There are more urgent issues to tackle regarding moderation until then, IMHO.

Chocobozzz commented:
Closing due to inactivity, feel free to reopen.

fflorent commented:
@kontrollanten Not only to rebase, but also:
- take into account the various remarks
- add the tests

fflorent commented:
@kontrollanten Also maybe update dependencies of this library:
https://github.com/rigelk/videojs-vtt-thumbnails
